### PR TITLE
Fix Issue 20964 - CTFE support for backward pointer iteration

### PR DESCRIPTION
**Pull request by 9il**

### CODE REVIEW

File: test/compilable/fix20964.d [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// https://issues.dlang.org/show_bug.cgi?id=20964
+
+uint foo(uint[3] m)
+{
+    auto p = m.ptr;
+    p += 3; // ok, p refers to the next right element after the block!
+    p -= 4; // ok, p refers to the next left element before the block!
```
Comment (thewilsonator) [inline test/compilable/fix20964.d:7]: this is missing the `++p;` from the test case.

**Review by thewilsonator — COMMENTED** [review]
Otherwise looks  OK to me, will defer to @UplinkCoder

**Comment by WalterBright** [comment]
I'm asking for a better rationale for this feature in the bugzilla.

File: test/compilable/fix20964.d (hunk 1 above)
Comment (9il) [reply test/compilable/fix20964.d:7]: Thanks, fixed

**Comment by 9il** [comment]
Assume `retro` is using iterators/pointers (it is required for Numpy like ndarrays)/

Take pseudo code `aSlice.retro.anotherOneLazyModificationFun.retro` as an example.

Walter's conventional solution with pre-decrement will work for the first retro, but it will lead to the same CTFE error for the second one `retro` in the pipeline chain.

File: src/dmd/ctfeexpr.d [hunk 2]
```diff
@@ -873,7 +873,9 @@ UnionExp pointerArithmetic(const ref Loc loc, TOK op, Type type, Expression eptr
         error(loc, "CTFE internal error: bad pointer operation");
         goto Lcant;
     }
-    if (indx < 0 || len < indx)
+    // We allow indx to be equal to the len: one position below the right bound.
+    // So we do for the left bound for backward iteration algorithms.
```
Comment (ibuclaw) [inline src/dmd/ctfeexpr.d:877]: Reverse iteration algorithms, because the keyword isn't `foreach_backwards`.

File: test/compilable/fix20964.d [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
```
Comment (ibuclaw) [inline test/compilable/fix20964.d:7]: Are there tests for indexing outside the bounds of the array that's being pointed at?

**Comment by ibuclaw** [comment]
```
auto idx = rng.length; // one past end
do
{
    --idx;
    // do something
} while (idx != 0)
```
Or
```
foreach (0 .. rng.length)
{
    --idx;
    // do something
}
```
Why wouldn't this work for ranges with non-contiguous memory layouts?

File: src/dmd/ctfeexpr.d [hunk 4]
```diff
@@ -873,7 +873,9 @@ UnionExp pointerArithmetic(const ref Loc loc, TOK op, Type type, Expression eptr
...
+    if (!(-1 <= indx && indx <= len))
```
Comment (UplinkCoder) [inline src/dmd/ctfeexpr.d:878]: compare to the previous code you reverse the order of the operands, changed the constant and added a negation.
anyone familiar with the previous code will be mighty surprised.
maybe there is a clearer way of stating this condition?

Comment (UplinkCoder) [reply src/dmd/ctfeexpr.d:878]:
```suggestion
(indx < -1 || len < indx)
```

**Comment by UplinkCoder** [comment]
so you want to be able to do:
```
auto p = &array[$-1];
auto end = &array[-1];
while(p != end)
{//iiterate}
```

I assume on the bases that `p != end`
feels faster than keeping a separate index variable around?

**Comment by UplinkCoder** [comment]
Ahh I see.
@9il  you are saying that it is fine for the pointer to point into garbage as long as it points to valid memory at the point where it is de-referenced.

Due to the way ctfe works we would need to separately track the changes to the pointer.
and condense that chain when we de-reference.

that is possible but requires more work than is currently in this PR.

**Comment by UplinkCoder** [comment]
@9il I am going to close this and implement it properly as and when I have time to do so.
If it is urgent work for a client I am sure we can work something out for me to prioritize this issue.

File: src/dmd/ctfeexpr.d (hunk 2 above)
Comment (9il) [reply src/dmd/ctfeexpr.d:877]: I don't understand you. What is the problem with `foreach_backwards`? It isn't related to the issue.

File: test/compilable/fix20964.d (hunk 3 above)
Comment (9il) [reply test/compilable/fix20964.d:7]: No idea. I have added the test for issue.

**Comment by 9il** [comment]
> Why wouldn't this work for ranges with non-contiguous memory layouts?

It works for Phobos ranges. It wouldn't work for Mir ranges. They have iterator based design to support multidimensional range API. The problem is that issue is blocker not for a particular loop but for the iterator based retro concept in general, see https://github.com/dlang/dmd/pull/13094#issuecomment-924642987.

> feels faster than keeping a separate index variable around?

For pointers it doesn't matter, but it is faster for a some kinds of lazy generic ranges. The problem is that we

**Comment by 9il** [comment]
> @9il you are saying that it is fine for the pointer to point into garbage as long as it points to valid memory at the point where it is de-referenced.
> 
> Due to the way ctfe works we would need to separately track the changes to the pointer.
> and condense that chain when we de-reference.
> 
> that is possible but requires more work than is currently in this PR.

It can point to -1 element of  the allocated range. We don't need the ability to for it to point anywhere. 

> @9il I am going to close this and implement it properly as and when I have time to do so. If it is urgent work for a client I am sure we can work something out for me to prioritize this issue.

It is important for a client. What is the proper way?

**Comment by UplinkCoder** [comment]
@9il the proper way it what I said before.
Rather than modifying the value of the pointer directly in the interpreter we keep a list of pending modifications and apply them when it's time to de-reference the pointer.

**Comment by UplinkCoder** [comment]
that way it's fine to temporarily invalidate the invariant of the having to point into valid memory, as long as it's restored at some point prior to use.

**Comment by 9il** [comment]
> Rather than modifying the value of the pointer directly in the interpreter we keep a list of pending modifications and apply them when it's time to de-reference the pointer.

That will kill the performance and will require huge memory allocations for some computation cases. We can make millions of pointers movements without dereferencing it at all.

What is wrong with the proposed solution? It just reflects existing logic for the left bound.

**Comment by UplinkCoder** [comment]
it's not general.
and it doesn't actually make your usecase work.

In that case we can introduce a field to the ptrExp that keeps track of the current modification to the base.
I.e. treat a pointer as a `size_t Base, ptrdiff_t Offset` struct.
and when it's time to deref we add the Offset and  Base.

**Comment by 9il** [comment]
> and it doesn't actually make your usecase work.

Could you reopen the MR? I will try to make it work at least.

**Comment by UplinkCoder** [comment]
I cannot because the branch was force-pushed to.

File: src/dmd/ctfeexpr.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/ctfeexpr.d:877]: It triggers semantic warnings for native English speakers. ;-)

Comment (9il) [reply src/dmd/ctfeexpr.d:877]: Got it:) thanks

**Pull request closed without merging**